File: Leetcode/2359 Find Closest Node to Given Two Nodes.java
class Solution {
    public int closestMeetingNode(int[] edges, int node1, int node2) {
        Integer[] dist1 = new Integer[edges.length];
        dfs(node1, -1, edges, dist1);

        Integer[] dist2 = new Integer[edges.length];
        dfs(node2, -1, edges, dist2);

        int minMaxDist = Integer.MAX_VALUE, minNode = -1;
        for (int node = 0; node < edges.length; ++node) {
            if (dist1[node] == null || dist2[node] == null) {
                continue;
            }
            int currMaxDist = Math.max(dist1[node], dist2[node]);
            if (currMaxDist < minMaxDist) {
                minMaxDist = currMaxDist;
                minNode = node;
            }
        }
        return minNode;
    }

    public void dfs(int node, int parent, int[] edges, Integer[] dist) {
        if (dist[node] != null) {
            return;
        }
        dist[node] = parent == -1 ? 0 : 1 + dist[parent];
        if (edges[node] >= 0) {
            dfs(edges[node], node, edges, dist);
        }
    }
}

File: Leetcode/2359 Find Closest Node to Given Two Nodes.py
class Solution:
    INF = float("inf")

    def closestMeetingNode(self, edges, node1, node2):
        dist1 = [Solution.INF for _ in edges]
        Solution.dfs(node1, None, edges, dist1)

        dist2 = [Solution.INF for _ in edges]
        Solution.dfs(node2, None, edges, dist2)

        min_max_dist, min_node = Solution.INF, -1
        for node, (d1, d2) in enumerate(zip(dist1, dist2)):
            cur_max_dist = max(d1, d2)
            if cur_max_dist < min_max_dist:
                min_max_dist = cur_max_dist
                min_node = node

        return min_node

    @staticmethod
    def dfs(node, parent, edges, dist):
        if node < 0 or dist[node] is not Solution.INF:
            return

        dist[node] = 0 if parent is None else 1 + dist[parent]
        Solution.dfs(edges[node], node, edges, dist)
